### PR TITLE
chore: Implement VedaAdapter

### PR DESCRIPTION
**Pull request by MoMannn**
### **What?**

- Adds `VedaAdapter.sol` — a new adapter contract for delegation-based deposits and withdrawals on Veda BoringVault
- Adds `IVedaTeller.sol` — minimal interface for Veda's TellerWithMultiAssetSupport
- Adds `VedaLending.t.sol` — comprehensive test suite (fork tests, happy paths, input validation, events, batch operations, edge cases, 3-level delegation chain, terms validation, replay prevention, slippage protection, alternative delegator)
- Adds `BatchDepositExecuted` and `BatchWithdrawExecuted` events for off-chain indexing of batch operations

### **Why?**

- Enables users to deposit into and withdraw from Veda BoringVaults through the delegation framework, without requiring direct token approvals
- Allows operators to execute vault operations on behalf of users via the redelegation pattern, with fine-grained control over tokens, amounts, and who can redeem

### **How?**

- Uses the redelegation pattern (user → operator → adapter) consistent with `DelegationMetaSwapAdapter`
- Parses deposit token/amount and withdrawal share amount from the leaf delegation's caveat terms via a shared `_parseERC20TransferTerms` helper, which validates that terms are at least 52 bytes (ERC20TransferAmountEnforcer format) before reading with assembly
- On deposit: redeems the delegation chain to transfer tokens to the adapter, approves the BoringVault, and calls the 5-arg `teller.deposit()` to mint shares directly to the root delegator
- On withdraw: redeems the delegation chain to transfer vault shares to the adapter, then calls `teller.withdraw()` to burn shares and send the desired underlying output asset to the root delegator. The output token is passed as a function parameter (distinct from the vault share token in the caveat)
- Supports batched operations via `depositByDelegationBatch` / `withdrawByDelegationBatch`, each emitting a batch-level event alongside per-stream `DepositExecuted` / `WithdrawExecuted` events
- Includes `withdrawEmergency` (owner-only) for recovering tokens accidentally sent to the contract

### **Note**

- Tests are being run on a mainnet fork of Arbitrum on a deployment of Veda protocol

---

> [!NOTE]
> **High Risk**
> New on-chain path moves user ERC20 and vault shares with permissionless `depositByDelegation` / `withdrawByDelegation`; safety depends on correct delegation caveats and Teller authorization, not caller identity.
> 
> **Overview**
> Adds **`VedaAdapter`**, a delegation-driven bridge to Veda **BoringVault** deposit and withdraw flows (same redeem-then-call pattern as the MetaSwap adapter). Operators redeem a **user → operator → adapter** chain via `DelegationManager`, pull **`depositToken`** or vault shares into the adapter, then call **`IVedaTeller`** so assets/shares land on the **root delegator**. Amounts come from **`caveats[0]`** (`ERC20TransferAmountEnforcer` terms, 52-byte check); batch entrypoints and owner **`withdrawEmergency`** / **`ensureAllowance`** are included.
> 
> Also adds **`DeployVedaAdapter`** (CREATE2 + env knobs), **`.env.example`** Veda vars, **chain 143 broadcast** artifacts, and **`VedaLending.t.sol`** (Arbitrum fork). CI tests gain **`ARBITRUM_RPC_URL`** and **`RPC_API_KEY`**.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit e2227eff0b03281d3b866a024143b52c89f27bd6. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/helpers/VedaAdapter.sol [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+
+        bytes[] memory executionCallDatas_ = new bytes[](1);
+        bytes memory encodedTransfer_ = abi.encodeCall(IERC20.transfer, (address(this), _amount));
+        executionCallDatas_[0] = ExecutionLib.encodeSingle(_token, 0, encodedTransfer_);
+
+        delegationManager.redeemDelegations(permissionContexts_, encodedModes_, executionCallDatas_);
+
+        // Approve BoringVault to pull tokens, then deposit via Teller
+        _ensureAllowance(IERC20(_token), boringVault, _amount);
+        uint256 shares_ = teller.deposit(_token, _amount, _minimumMint, rootDelegator_, address(0));
```
[the remark this answers is not in the document]
Comment (MoMannn) [reply src/helpers/VedaAdapter.sol]: intended

File: src/helpers/VedaAdapter.sol [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+     * @param _shareAmount Amount of vault shares to redeem
+     * @param _minimumAssets Minimum underlying assets the user expects to receive (slippage protection)
+     * @notice Security consideration: Callable by anyone. The redelegation passed in MUST include an
+     *      `ERC20TransferAmountEnforcer` capped to exactly `_shareAmount` to prevent over-spending or replay.
+     */
+    function withdrawByDelegation(
+        Delegation[] memory _delegations,
+        address _token,
+        uint256 _shareAmount,
+        uint256 _minimumAssets
```
[the remark this answers is not in the document]
Comment (MoMannn) [reply src/helpers/VedaAdapter.sol]: Not true. The delegation itself enforces which token and amount can be withdrawn. The input itself if just for simpler API and so that we don't need to gather that from the delegation.

File: script/DeployVedaAdapter.s.sol [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ * @notice Deploys the VedaAdapter contract.
+ * @dev Update the hardcoded addresses below before deploying.
+ * @dev Fill the SALT variable in the .env file
+ * @dev run the script with:
+ * forge script script/DeployVedaAdapter.s.sol --rpc-url <your_rpc_url> --private-key $PRIVATE_KEY --broadcast
+ */
+contract DeployVedaAdapter is Script {
+    // Hardcoded constructor parameters - update these before deploying
+    address constant OWNER = address(0x0000000000000000000000000000000000000000);
+    address constant DELEGATION_MANAGER = address(0x0000000000000000000000000000000000000000);
```
Comment (McOso) [inline script/DeployVedaAdapter.s.sol]: lets move these to `vm.env..` calls, like else where in repo.

File: src/helpers/interfaces/IVedaTeller.sol [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        uint256 depositAmount,
+        uint256 minimumMint,
+        address referralAddress
+    )
+        external
+        payable
+        returns (uint256 shares);
+
+    /**
+     * @notice Allows an authorized caller to deposit into the BoringVault for another address, if this contract is not paused.
```
Comment (McOso) [inline src/helpers/interfaces/IVedaTeller.sol:36]: nit: add @param like other functions

File: src/helpers/VedaAdapter.sol [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+     * @param _minimumMint Minimum vault shares the caller expects to receive, used as a sanity-check
+     *      bound. The Veda vault conversion is always at fair value; rate drift from yield streaming
+     *      is negligible. A tolerance of 0.1-0.5% is recommended. If this check causes a revert,
+     *      no funds are lost — retry with a fresh quote.
+     * @notice Security consideration: Callable by anyone. The redelegation passed in MUST include an
+     *      `ERC20TransferAmountEnforcer` capped to exactly the intended deposit amount to prevent
+     *      over-spending or replay.
+     */
+    function depositByDelegation(Delegation[] memory _delegations, uint256 _minimumMint) external {
+        _executeDepositByDelegation(_delegations, _minimumMint, msg.sender);
```
Comment (McOso) [inline src/helpers/VedaAdapter.sol]: you dont need to pass in `msg.sender` as a param. `msg.sender` does not change in the internal function

File: src/helpers/VedaAdapter.sol [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+     *      from the first caveat of each stream's leaf delegation.
+     * @param _depositStreams Array of deposit parameters
+     * @notice Security consideration: Callable by anyone. Each redelegation in the batch MUST include an
+     *      `ERC20TransferAmountEnforcer` capped to exactly the intended deposit amount to prevent over-spending or replay.
+     */
+    function depositByDelegationBatch(DepositParams[] memory _depositStreams) external {
+        uint256 streamsLength_ = _depositStreams.length;
+        if (streamsLength_ == 0) revert InvalidBatchLength();
+
+        address caller_ = msg.sender;
```
Comment (McOso) [inline src/helpers/VedaAdapter.sol]: adds just a hair more gas to cache it

File: src/helpers/VedaAdapter.sol [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+    /**
+     * @notice Emergency function to recover tokens accidentally sent to this contract
+     * @dev This contract should never hold ERC20 tokens as all token operations are handled
+     *      through delegation-based transfers that move tokens directly between users and the BoringVault.
+     *      This function is only for recovering tokens sent to this contract by mistake.
+     * @param _token The token to be recovered
+     * @param _amount The amount of tokens to recover
+     * @param _recipient The address to receive the recovered tokens
+     */
+    function withdrawEmergency(IERC20 _token, uint256 _amount, address _recipient) external onlyOwner {
```
Comment (McOso) [inline src/helpers/VedaAdapter.sol:302]: hmm im thinking we get rid of this function. It doesnt seem possible for tokens to be left in this contract without the tx reverting, since it's atomic. If we remove, it really makes the adapter clean and lightweight, removes the need for owner / state.

File: src/helpers/VedaAdapter.sol [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+    /**
+     * @notice Ensures sufficient token allowance for a spender to pull tokens
+     * @dev Checks current allowance and sets exact amount if insufficient, avoiding accumulation
+     * @param _token Token to manage allowance for
+     * @param _spender Address that needs to spend the tokens
+     * @param _amount Amount needed for the operation
+     */
+    function _ensureAllowance(IERC20 _token, address _spender, uint256 _amount) private {
+        uint256 allowance_ = _token.allowance(address(this), _spender);
+        if (allowance_ < _amount) {
```
Comment (McOso) [inline src/helpers/VedaAdapter.sol]: why not just set infinite approval amount here? The actual amount of token transfer is guarded by the delegation

File: src/helpers/VedaAdapter.sol [hunk 9]
```diff
@@ -0,0 +1,402 @@
...
+     *      via `_parseERC20TransferTerms`.
+     * @param _delegations Delegation chain, sorted leaf to root
+     * @param _minimumMint Minimum vault shares expected (sanity-check bound)
+     * @param _caller Address of the caller, used only for event emission
+     */
+    function _executeDepositByDelegation(Delegation[] memory _delegations, uint256 _minimumMint, address _caller) internal {
+        uint256 length_ = _delegations.length;
+        if (length_ < 2) revert InvalidDelegationsLength();
+
+        (address token_, uint256 amount_) = _parseERC20TransferTerms(_delegations[0].caveats[0].terms);
```
Comment (hanzel98) [inline src/helpers/VedaAdapter.sol]: The docs might explain it but nothing enforces that the first caveat is erc20 transfer enforcer.

File: src/helpers/VedaAdapter.sol [hunk 10]
```diff
@@ -0,0 +1,402 @@
...
+
+    /**
+     * @notice Internal implementation of deposit by delegation
+     * @dev Parses the deposit token and amount from the first caveat of the leaf delegation
+     *      via `_parseERC20TransferTerms`.
+     * @param _delegations Delegation chain, sorted leaf to root
+     * @param _minimumMint Minimum vault shares expected (sanity-check bound)
+     * @param _caller Address of the caller, used only for event emission
+     */
+    function _executeDepositByDelegation(Delegation[] memory _delegations, uint256 _minimumMint, address _caller) internal {
```
Comment (hanzel98) [inline src/helpers/VedaAdapter.sol]: It seems like native token transfers don't make sense through this adapter since they don't need the approval, but be sure to think about any possible security issues of the native token flow

File: src/helpers/VedaAdapter.sol (hunk 7 above)
Comment (MoMannn) [reply src/helpers/VedaAdapter.sol:302]: This is in case of someone sending tokens directly to the contract (mistaking it for EOA).

File: src/helpers/VedaAdapter.sol (hunk 9 above)
Comment (MoMannn) [reply src/helpers/VedaAdapter.sol]: Its not directly enforced on the contract but added docs that it should be like that. Basically part of the usage protocol. If we set it directly to an address we would need adapter and new delegations in the case of upgrade

**Review by McOso — APPROVED** [review]
